Various documentation updates

QuLogic opened the pull request:
## PR Summary

These are some minor preemptive corrections to the docs.

Firstly, import the unreleased What's New? entries into the docs so we can see what the output is like. There was one bug in there that I fixed. Unfortunately, Sphinx cannot see both a `whats_new.rst` file and a `whats_new/` directory, so I had to rename one. I hope the instructions are reasonable for a release step @tacaswell.

Secondly, due to sphinx-gallery/sphinx-gallery#250, several things that were just printed out are now real Sphinx warnings, which will fail with our default build setup. I corrected the back-reference URLs so that sphinx-gallery stops warning and creates them now. Unfortunately, this does slow the build a bit, but I'm hoping sphinx-gallery/sphinx-gallery#272 can be merged as well to help with that.

Finally, with the deploy key added on CircleCI and devdocs, enable the new key so we can push the build docs on `master`. This should fix #8875. I have tested this on [my fork](https://github.com/matplotlib/matplotlib/compare/master...QuLogic:master) and CircleCI correctly [uploads docs](https://circleci.com/gh/QuLogic/matplotlib/70) once this change is made.

## PR Checklist

- [N/A] Has Pytest style unit tests
- [N/A] Code is PEP 8 compliant 
- [N/A] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [N/A] Added an entry to doc/users/whats_new.rst if major new feature
- [N/A] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

QuLogic commented on doc/users/whatsnew/multiple_legend_keys.rst:
```diff
@@ -6,11 +6,9 @@ A legend entry can now contain more than one legend key. The extended
 
-.. figure:: ../../gallery/pylab_examples/images/sphx_glr_legend_demo6_001.png
-   :target: ../../gallery/pylab_examples/legend_demo6.html
+.. figure:: /gallery/text_labels_and_annotations/images/sphx_glr_legend_demo_004.png
```
PS, @choldgraf for these, we don't need the `../..` because Sphinx accepts an absolute path. Unfortunately, it doesn't work on the target below though.

choldgraf replied:
that's good to know! thanks

QuLogic commented:
Oops, I forgot the currently released version of sphinx-gallery doesn't support http:// links properly, so I reverted that bit. Hopefully should build now.

choldgraf commented on doc/devel/release_guide.rst:
```diff
@@ -48,7 +48,10 @@ Review and commit changes.  Some issue/PR titles may not be valid rst (the most
 ----------
 
-Before tagging, make sure that the docs build cleanly ::
+Before tagging, update the what's new listing in :file:`doc/users/whats_new.rst`
```
I don't think it should be done in this PR, but I still think it'd be good to have some kind of "credit" system for various features. E.g. just making sure that every new feature or bugfix has a `by XXX` at the end, optionally with a link to a user webpage or something.

choldgraf commented:
Almost all of this LGTM. The one thing I'm a little concerned about is having a `whats_new.rst` and a `whatsnew` folder. Why not call the folder `whats_new_cache` or something like that...so that it's clear it's a temporary holding place for new content that hasn't been incorporated into the main `whats_new.rst` file yet?

QuLogic commented:
I renamed it to `next_whats_new` to mirror the existing `prev_whats_new` directory. Also, reverted the change from https to http above as the latest release of sphinx-gallery is out now and supports https. With the new sphinx-gallery release, warnings are true warnings, so we do need to merge this soon-ish to fix that.

tacaswell commented:
:+1: doing that with the whats_new folder has been on my list of things to get to for a while!

choldgraf commented:
Then I saw let's do it...tests are passing and I think the new folder naming works great

choldgraf commented:
Thanks @QuLogic !!! 🎉

QuLogic commented:
Fingers crossed on the deploy this time...
